### ISSUE TITLE
[Feature Request] Local Scraper (Use browser auth)

### ISSUE THREAD

**Issue by brucealdridge** [hoarder-app/hoarder#172]
There are a number of sites that I visit that I would like to bookmark that require authentication.  A good example of this is news sites with content behind paywalls.

Using a scraper on a server won't work and will instead just save a login screen.

Can the browser extensions pass a copy of the page via the API and save that?

**Comment by kamtschatka** [comment]
Yes, taking screenshots is possible with chrome extensions.
One issue would be that rescraping the page would not be possible as you would go to the login screen again, so there would need to be some kind of "prevent rescraping" flag.
Another option would probably be to use your locally installed chrome instance for scraping the data by running a worker locally. I am not sure how user friendly that would be^^.

**Comment by MohamedBassem** [comment]
This makes a lot of sense. The extension itself can capture the page content so that hoarder doesn't need to crawl it. This is a reasonable feature request, will add it to our todo list :)

**Comment by kureta** [comment]
This would be a great feature. Also [tubearchivist](https://github.com/tubearchivist/browser-extensionl) has a browser extension that syncs your youtube cookies with the tube archivist server. An extension that automatically shares all your cookies, or let's you choose which cookies to share, or sends the cookies of current page to hoarder before it starts scraping might be an option.

**Comment by javydekoning** [comment]
A similar solution to Evernote Web clipper would be awesome. 

Select some text/images -> right click -> hoard.

https://chromewebstore.google.com/detail/evernote-web-clipper/pioclpoplcdbaefihamjohnefbikjilc?hl=en

**Comment by huyz** [comment]
See also https://github.com/webclipper/web-clipper

**Comment by NotChristianGarcia** [comment]
^ web-clipper does exist and work. I didn't like the flow too much.

[SingleFile](https://github.com/gildas-lormeau/SingleFile) is another project to check. It outputs .html (or an archive). I think it's easier to manage and quick to run. It has an `Upload to a REST form API` option in settings that sets a destination that hoarder could use if hoarder doesn't want to re-implement scraping.

**Comment by errorsandwarnings** [comment]
@NotChristianGarcia  I agree, and came looking for similar functionality. All it needs is a REST API endpoint which single file can send to and the API endpoint parses the archive and puts it in hoarder. This will help solve so many problems. Local scrape is a really needed feature.

**Comment by yinan-c** [comment]
That's what I have in mind as well.

**Comment by MohamedBassem** [comment]
Hey folks, I know how important this feature is and it's on top of my todo list! I'll see if I can get it in the next release!

**Comment by freakshock88** [comment]
I tried uploading a hosted HTML file that I downloaded with singlefile. Hoarder does not seem to be able to process these HTML archive files. Did anyone try that before?

**Comment by kamtschatka** [comment]
if someone else has tried it, they would get the same result, as this is not supported

**Comment by freakshock88** [comment]
Actually, I was wrong. Hoarder was able to process my selfhosted HTML archive files generated by Singlefile. It just took more time than I expected. I see a few timeouts in the logs for the Hoarder worker, but after like 10 minutes the page has been successfully added to hoarder.

**Comment by kamtschatka** [comment]
what exactly are you doing? I don't understand what the worker has to do with you adding a html file? how are you even adding the html file?

**Comment by agruen** [comment]
@MohamedBassem  -- I just wanted to add that if you offered a way to pass content of a page through the API, there are opportunities to make this work with Mac/iOS shortcuts.  (Which is particularly important for Safari on Mac users... as that's how we can currently save!)

**Comment by MohamedBassem** [comment]
So folks, as a stopgap until this is supported natively in hoarder's extension, I've just published 10506173cd5309e7c63d83055243abc67cecad4f which exposes an API from hoarder that's compatible with the singleFile extension directly.

It should be available in the nightly build in ~15mins. If you move to the nightly release track, you can use it by following this [guide](https://docs.hoarder.app/next/Guides/singlefile). Please give it a try and let me know how it goes.

Keep in mind that once you go to the nightly track, it's not safe to go back to the stable release until the next release.

**Comment by freakshock88** [comment]
> what exactly are you doing? I don't understand what the worker has to do with you adding a html file? how are you even adding the html file?

I am downloading a singlefile generated html file, and then hosting that file on my own domain. Then I send that url to hoarder. I first thought hoarder was incompatible with the singlefile generated html files, but that is not the case. But @MohamedBassem's solution will be way easier :)

**Comment by huyz** [comment]
> So folks, as a stopgap until this is supported natively in hoarder's extension, I've just published https://github.com/hoarder-app/hoarder/commit/10506173cd5309e7c63d83055243abc67cecad4f which exposes an API from hoarder that's compatible with the singleFile extension directly.

Gamechanger alert! 😄

**Comment by dionorgua** [comment]
Wow. Thanks!. This is huge improvement.

**Comment by errorsandwarnings** [comment]
@MohamedBassem  When is the next stable planned to release?

**Comment by MohamedBassem** [comment]
@errorsandwarnings We've had a release last week, so I'd expect the next release to be at least a couple of weeks away unfortunately.

**Comment by errorsandwarnings** [comment]
@MohamedBassem Is it safe to switch to nightly and then to the next stable when released? Thinking of avoiding loosing any data or links.

**Comment by errorsandwarnings** [comment]
@MohamedBassem  Just tried it, it can archive small pages but it started hitting too big to store limit on some pages.

**Comment by brucealdridge** [comment]
> @MohamedBassem Just tried it, it can archive small pages but it started hitting too big to store limit on some pages.

[Default is set to 4MB, looks like its set by ENV vars.](https://github.com/hoarder-app/hoarder/blob/bbbf33536c6c1a73fdfe76c42ff19cd08aabcc6f/docs/docs/03-configuration.md?plain=1#L12)

**Comment by MohamedBassem** [comment]
@errorsandwarnings yes, you can go to nightly and back to stable in the next release.

@brucealdridge Yeah, I probably should change the default to 50MB. 4MB is going to cause a lot of hassle to a lot of people using this feature.

**Comment by errorsandwarnings** [comment]
@MohamedBassem After changing the limit, it starting working. However I do see there is quite a lot of lag, the link in hoarder shows up but the actual archive file is taking 5-10 minutes to populate and become visible.

**Comment by MohamedBassem** [comment]
please share the logs if possible :)

**Comment by errorsandwarnings** [comment]
@MohamedBassem  Log location?

**Comment by MohamedBassem** [comment]
just from 'docker compose logs', if you're using docker

**Comment by errorsandwarnings** [comment]
@MohamedBassem  Try this : http://0x0.st/8-Ff.txt

**Comment by errorsandwarnings** [comment]
The reddit archive I have tried to upload never shows up. I think the other urls are likely being scraped and not coming from singlefile upload.

**Comment by ironicbadger** [comment]
Singlefile gives the following error when attempting to save to Hoarder via the nightly release.

<img width="273" alt="Image" src="https://github.com/user-attachments/assets/85e40044-f15b-4b51-9fe2-223f02a14665" />

Followed these instructions for configuration.
https://docs.hoarder.app/next/Guides/singlefile/

**Comment by DarthDobber** [comment]
@ironicbadger I had that same error, in my case I haven't yet setup SSL for my hoarder instance so I had the change it from https (as it says in the instructions) to http and then it started working.

**Comment by DarthDobber** [comment]
![Image](https://github.com/user-attachments/assets/9970596e-58df-4880-bd64-c3dd1778f143)

I am getting this error when attempting to hoard this URL

https://www.msn.com/en-us/news/us/ar-AA1xovTT

**Comment by ironicbadger** [comment]
> @ironicbadger I had that same error, in my case I haven't yet setup SSL for my hoarder instance so I had the change it from https (as it says in the instructions) to http and then it started working.
> 
> 

i tried both http to the IP or my reverse proxy https url as well

**Comment by psjay** [comment]
handy feature! a small suggestion here: hoarder could automatically enable the archive version of the uploaded singfile since it's already an archive.

**Comment by dwayne-hu** [comment]
> ![Image](https://github.com/user-attachments/assets/9970596e-58df-4880-bd64-c3dd1778f143)
> 
> I am getting this error when attempting to hoard this URL
> 
> https://www.msn.com/en-us/news/us/ar-AA1xovTT

I am having the same problem, despite i already set the maximum asset size to 50 MB. Does anybody have a solution?

**Comment by errorsandwarnings** [comment]
Tried multiple times, can't get this working properly. The link is created in hoarder but the sent singlefile archive is never seen. @MohamedBassem  Any reasons? I did attach docker logs earlier.

**Comment by DarthDobber** [comment]
> I am having the same problem, despite i already set the maximum asset size to 50 MB. Does anybody have a solution?
> 
> Edit: I checked the file size of the created singlefile. It is about 160MB for some reddit sites. This problem is adressed in this [issue](https://github.com/gildas-lormeau/SingleFile/issues/1387). My problem is solved when enabling `Stylesheets > group duplicate stylesheets together`

Thank you for this!  I made the stylesheet config change and it helped with some articles.  However, not all.  For example, this url still returns the error:

https://www.thefp.com/p/dont-free-ross-ulbricht-clemency-trump

**Comment by jasonyang-ee** [comment]
> Singlefile gives the following error when attempting to save to Hoarder via the nightly release.
> <img alt="Image" width="273" src="https://private-user-images.githubusercontent.com/2773080/404358152-85e40044-f15b-4b51-9fe2-223f02a14665.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgzMTI3NDUsIm5iZiI6MTczODMxMjQ0NSwicGF0aCI6Ii8yNzczMDgwLzQwNDM1ODE1Mi04NWU0MDA0NC1mMTViLTRiNTEtOWZlMi0yMjNmMDJhMTQ2NjUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEzMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMzFUMDgzNDA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjkxNDQ0YTNjNzg2MDc4ODZlZTY2ZTVmZmVkMDZlNTk1ZTBlOGJjN2Y4MmZjODY0MjU5ZDE1NTUwNGQwMzBkYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.k8aJQU1YGJP8o4iFagcvXgghaL61zBKaBrqCVdGMlaA">
> 
> Followed these instructions for configuration. https://docs.hoarder.app/next/Guides/singlefile/

I have the same issue. I feel this is due to the update is still in "next" branch which is not yet released.

```
This is unreleased documentation for Hoarder Docs Next version.
```

**Comment by huyz** [comment]
TIPs:

In SingleFile options
- You can make use of profiles to have one set to direct to Hoarder.
- Make sure that the File Format is set to HTML and not Zip file.

**Comment by ctschach** [comment]
So I tried to setup SingleFile on iPadOS and it doesn‘t work. I mean I can get SingleFile to work inside the browser, but using the REST API and push this to hoarder does not work.

The strange thing is: I didn‘t even get an error message. So this file is just vacuumed into a black hole…

I increased the upload limit to 200MB, so this should be enough. 

The is the page I try to save:

https://www.etsy.com/de/listing/1702261359/fiat-ducato-kleiderhaken-the-hooks

**Comment by errorsandwarnings** [comment]
The same is happening to me, it goes somewhere, but where I don't know.

**Comment by errorsandwarnings** [comment]
Just used the latest version and it seems to be working now.

**Comment by errorsandwarnings** [comment]
@brucealdridge  Does the search work inside the archive ? Also, I don't see it picking title correctly

**Comment by brucealdridge** [comment]
> @brucealdridge  Does the search work inside the archive ? Also, I don't see it picking title correctly

If something isn't working open a new issue. It helps if you have clear reproduction steps.

**Comment by bbrendon** [comment]
If anyone is wondering why this was closed. I believe its this new feature : See https://docs.hoarder.app/next/Guides/singlefile/

**Comment by AlejandroAkbal** [comment]
I think this should be reopened and implement single file natively into Karakeep extension, as realistically no one is going to learn about single file extension, install it and configure it

Similar to how Wallabag does it natively in their extension

**Comment by huyz** [comment]
> realistically no one is going to learn about single file extension, install it and configure it

That isn't even remotely realistic.  The Singlefile extension is hugely popular and has other purposes besides Karakeep.

**Comment by AlejandroAkbal** [comment]
Talking from my experience as a developer and common user; I think less than 10% of Karakeep users are going to handle this case by installing another extension and configuring it, while I do believe many users would use it if it was natively integrated

Has nothing to do with singlefile's popularity and more with the laziness of using and maintaining yet another extension for a specific use case
